service/header/sync.go

```diff
@@ -330,3 +427,120 @@ func (s *Syncer) getHeaders(ctx context.Context, start, amount uint64) ([]*Exten
 
 	return out, nil
 }
+
+// heightSub provides a way to wait until a specific height becomes available and synced
+type heightSub struct {
+	// usually we don't attach context to structs, but this is an exception
+	// as GetByHeight and ProvideHeights needs to access it
+	ctx       context.Context
+	cancel    context.CancelFunc
+	store     Store
+	height    uint64
+	provideCh chan []*ExtendedHeader
+	reqsCh    chan *heightReq
+	reqs      map[uint64][]chan *heightResp
+}
+
+type heightReq struct {
+	resp   chan *heightResp
+	height uint64
+}
+
+type heightResp struct {
+	header *ExtendedHeader
+	err    error
+}
+
+func newHeightSub(store Store) *heightSub {
+	return &heightSub{
+		store:     store,
+		provideCh: make(chan []*ExtendedHeader, 4),
+		reqsCh:    make(chan *heightReq, 4),
+		reqs:      make(map[uint64][]chan *heightResp),
+	}
+}
+
+func (hs *heightSub) Start() {
+	hs.ctx, hs.cancel = context.WithCancel(context.Background())
+	go hs.subLoop()
+}
+
+func (hs *heightSub) Stop() {
+	hs.cancel()
+}
+
+func (hs *heightSub) GetByHeight(ctx context.Context, height uint64) (*ExtendedHeader, error) {
+	h, err := hs.store.GetByHeight(ctx, height)
+	if err != ErrNotFound {
+		return h, err
+	}
+
+	resp := make(chan *heightResp, 1)
+	select {
+	case hs.reqsCh <- &heightReq{resp, height}:
+		select {
+		case resp := <-resp:
+			return resp.header, resp.err
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-hs.ctx.Done():
+			return nil, hs.ctx.Err()
+		}
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-hs.ctx.Done():
+		return nil, hs.ctx.Err()
+	}
+}
+
+func (hs *heightSub) ProvideHeights(ctx context.Context, headers ...*ExtendedHeader) {
```

Comment:
> So you mean for this method to be accessed outside the scope of this pkg by DASer I'm assuming

No, otherwise I would make it public. Think of heightSub as a subcomponent of Syncer. Syncer is anyway the main object regarding sync logic and heightSub is part of it, but I extracted it into a separate subcomponent, not to bloat Syncer. I might also extract it into `sync_heightsub.go` file.

So, no interface is needed and you can directly access `Syncer.GetByHeight`(which calls `hieghtSub.GetByHeight` underneath) to wait or get the Header instantly.
> but we need an interface for this since the struct itself is private.

The alternative to this would be making `heightSub` a subcomponent of Store. That is, the `Store.GetByHeight` would block until the Header is found with `heightSub.GetByHeight` underneath instead of erroring if not found. Also `heightSub.ProvideHeights` would be called by `Store.Append` when new headers are appended. This could also simplify heightSub a bit and remove the deadlock case.

Reply:
> No, otherwise I would make it public. Think of heightSub as a subcomponent of Syncer. Syncer is anyway the main object regarding sync logic and heightSub is part of it, but I extracted it into a separate subcomponent, not to bloat Syncer. I might also extract it into sync_heightsub.go file.

SGTM - I like the idea of extracting it but I also appreciate you maintaining it separately to prevent bloat

> So, no interface is needed and you can directly access Syncer.GetByHeight(which calls hieghtSub.GetByHeight underneath) to wait or get the Header instantly.

So DASer will need to take a full syncer then instead of a subscription?

> The alternative to this would be making heightSub a subcomponent of Store. That is, the Store.GetByHeight would block until the Header is found with heightSub.GetByHeight underneath instead of erroring if not found. Also heightSub.ProvideHeights would be called by Store.Append when new headers are appended. This could also simplify heightSub a bit and remove the deadlock case.

If this leads to simplification, I'm for it.

Reply:
Yes, it might take the whole syncer to access its GetByHeight, unless we do not go with lowering the heightSub down to the Store. In that case, DASer will only access Store and IMO it is better than passing Syncer there. Note that there won't be any subscription structs for a height specifically.
> So DASer will need to take a full syncer then instead of a subscription?

Reply:
Well, I am thinking now of extracting this into separate PR

Reply:
Done https://github.com/celestiaorg/celestia-node/pull/428

Reply:
One more related PR incoming.

Reply:
An uncontrollable male urge to refactor and optimize the shit our of the Store.

Reply:
resolved?

Reply:
Yes and the PR is merged.